core/chains/evm/abi/selector_parser.go

```diff
@@ -0,0 +1,249 @@
+	if unescapedSelector[0] == '(' {
+		return parseCompositeType(unescapedSelector)
+	}
+	return parseElementaryType(unescapedSelector)
+}
+
+func parseArgs(unescapedSelector string) ([]abi.ArgumentMarshaling, error) {
+	if len(unescapedSelector) == 0 || unescapedSelector[0] != '(' {
+		return nil, fmt.Errorf("expected '(', got %c", unescapedSelector[0])
+	}
+	result := []abi.ArgumentMarshaling{}
+	rest := unescapedSelector[1:]
+	var parsedType any
+	var err error
+	for len(rest) > 0 && rest[0] != ')' {
+		// parse method name
+		var name string
+		name, rest, err = parseIdentifier(rest[:])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse name: %v", err)
+		}
+
+		// skip whitespace between name and identifier
+		for rest[0] == ' ' {
+			rest = rest[1:]
+		}
+
+		// parse type
+		parsedType, rest, err = parseType(rest[:])
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse type: %v", err)
+		}
+
+		arg, err := assembleArg(name, parsedType)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse type: %v", err)
+		}
+
+		result = append(result, arg)
+
+		for rest[0] == ' ' || rest[0] == ',' {
+			rest = rest[1:]
+		}
+	}
+	if len(rest) == 0 || rest[0] != ')' {
+		return nil, fmt.Errorf("expected ')', got '%s'", rest)
+	}
+	if len(rest) > 1 {
+		return nil, fmt.Errorf("failed to parse selector '%s': unexpected string '%s'", unescapedSelector, rest)
+	}
+	return result, nil
+}
+
+func assembleArg(name string, arg any) (abi.ArgumentMarshaling, error) {
+	if s, ok := arg.(string); ok {
+		return abi.ArgumentMarshaling{Name: name, Type: s, InternalType: s, Components: nil, Indexed: false}, nil
+	} else if components, ok := arg.([]interface{}); ok {
+		subArgs, err := assembleArgs(components)
+		if err != nil {
+			return abi.ArgumentMarshaling{}, fmt.Errorf("failed to assemble components: %v", err)
+		}
+		tupleType := "tuple"
+		if len(subArgs) != 0 && subArgs[len(subArgs)-1].Type == "[]" {
+			subArgs = subArgs[:len(subArgs)-1]
+			tupleType = "tuple[]"
+		}
+		return abi.ArgumentMarshaling{Name: name, Type: tupleType, InternalType: tupleType, Components: subArgs, Indexed: false}, nil
+	}
+	return abi.ArgumentMarshaling{}, fmt.Errorf("failed to assemble args: unexpected type %T", arg)
+}
+
+func assembleArgs(args []interface{}) ([]abi.ArgumentMarshaling, error) {
+	arguments := make([]abi.ArgumentMarshaling, 0)
+	for i, arg := range args {
+		// generate dummy name to avoid unmarshal issues
+		name := fmt.Sprintf("name%d", i)
+		arg, err := assembleArg(name, arg)
+		if err != nil {
+			return nil, err
+		}
+		arguments = append(arguments, arg)
+	}
+	return arguments, nil
+}
+
+// ParseSelector converts a method selector into a struct that can be JSON encoded
+// and consumed by other functions in this package.
+// Note, although uppercase letters are not part of the ABI spec, this function
+// still accepts it as the general format is valid.
+func ParseSelector(unescapedSelector string) (abi.SelectorMarshaling, error) {
```

Comment:
What's the difference between ParseSelector and ParseSignature (below)? Seems like ParseSelector is not used anywhere in core.

Reply:
@DeividasK , @austinborn these two files were copied from geth (see comment above: // Sourced from https://github.com/ethereum/go-ethereum/blob/fe91d476ba3e29316b6dc99b6efd4a571481d888/accounts/abi/selector_parser.go#L126
// Modified assembleArgs to retain argument names)

Blaz modified them slightly to include argument name parsing as geth by default doesn't support that. They happened to be merged with my PR because encoders also depend on this library and were merged earlier.

We can either keep them as close to original as possible or make cleaner/smaller (I already had to do minor tweaks to satisfy Sonar). I'll leave that decision to @archseer and @patrick-dowell .